app/src/main/java/org/astraea/cost/HasBrokerCost.java

```diff
@@ -7,5 +7,5 @@ public interface HasBrokerCost extends CostFunction {
    * @param clusterInfo cluster information
    * @return the score of each broker. The score ranges in [0 - 1].
    */
-  BrokerCost brokerCost(ClusterInfo clusterInfo);
+  BrokerCost brokerCost(ClusterInfo clusterInfo, Normalizer normalizer);
```

Comment:
這裡有一個設計可以討論一下，就是`Normalizer`要放在哪一層，有兩個選擇：

1. 放在`HasBrokerCost`這一層
2. 放在`BrokerCost`這一層

我個人是偏好第二個的原因是先前 @chinghongfang  把`BrokerCost`這個邏輯抽出來就是要為之後做“多樣化”的查詢方式鋪路，也就是說任意一個cost function只管輸出“原始”資料並組成`BrokerCost`，而`BrokerCost`可以套用各種不同"演算法"來形成各種有意義的值

現在這個放在`HasBrokerCost`的話會導致輸出的第一站就綁死要用`normalization`，這樣就顯得侷限，因為有可能之後某個人寫的dispatcher要整理數據的第一步並不是`Normalization`

@chinghongfang @garyparrot @qoo332001 你們的想法呢

Reply:
確實放在HasBrokerCost裏用起來有些彆扭。
```
public interface BrokerCost {
  static BrokerCost normalize(Normalizer normalizer){
    return map->{
      var normalization = normalizer.normalize(map.values()).iterator();
      map.replaceAll((broker, v) -> normalization.next());
      return map;
    };
  }

  /** @return broker-id and its "cost" */
  Map<Integer, Double> value(Map<Integer, Double> map);
}
```
所以BrokerCost是改爲像是這樣嗎。

Reply:
> 所以BrokerCost是改爲像是這樣嗎。

可否給一段範例來說明你“希望”使用者會寫出什麼樣的程式碼？

Reply:
`return ignore->BrokerCost.normalize(Normalizer.TScore()).value(costMetrics);`
這能稱的上是範例嗎

Reply:
> return ignore->BrokerCost.normalize(Normalizer.TScore()).value(costMetrics);

這段程式碼看起來是各個Cost function要使用？還是使用cost function的使用者要呼叫？

Reply:
是各個Cost function要使用這段程式碼（會被賽到`brokerCost()`中）

Reply:
> 是各個Cost function要使用這段程式碼（會被賽到brokerCost()中）

那要先確認一件事情，現在cost function是否要回傳normalized values? I assume that cost function return "raw" cost instead of normalized cost, since the `BrokerCost` can be normalized later (if needs).

Reply:
學長是在說`HasBrokerCost`中的`brokerkerCost(ClusterInfo clusterInfo)`方法對嗎，按照其註解，當時的想法是返回一個0-1的數，這代表一定要經過歸一化。
/**
   * score all nodes according to passed beans and cluster information.
   *
   * @param clusterInfo cluster information
   * @return the score of each broker. The score ranges in [0 - 1].
   */

Reply:
> @return the score of each broker. The score ranges in [0 - 1].

這是可以調整的，就我前面 (https://github.com/skiptests/astraea/pull/389#discussion_r889010066) 提到，成本算出來後要不要正規化應該要讓後面接手的“演算法”來決定，而不是一開始就直接綁死一定要輸出正規化的值。

Reply:
噢噢 我瞭解了那我會把這一邏輯放到後面的演算法，而不是直接在`brokerCost()`回傳時進行normalize處理。

Reply:
我修改了一下，他現在大致變成了這樣。算法的邏輯修改我會留到下一個pr，這樣這支就不會很臃腫。

Reply:
> 我修改了一下，他現在大致變成了這樣。算法的邏輯修改我會留到下一個pr，這樣這支就不會很臃腫。

先開個議題記錄一下

Reply:
添加至 #393